controllers/effective_tls_policies_reconciler.go

```diff
@@ -0,0 +1,289 @@
+package controllers
+
+import (
+	"context"
+	"reflect"
+	"sync"
+
+	certmanv1 "github.com/cert-manager/cert-manager/pkg/apis/certmanager/v1"
+	"github.com/kuadrant/policy-machinery/controller"
+	"github.com/kuadrant/policy-machinery/machinery"
+	"github.com/samber/lo"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/client-go/dynamic"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	crlog "sigs.k8s.io/controller-runtime/pkg/log"
+	gatewayapiv1 "sigs.k8s.io/gateway-api/apis/v1"
+
+	kuadrantv1alpha1 "github.com/kuadrant/kuadrant-operator/api/v1alpha1"
+)
+
+type EffectiveTLSPoliciesReconciler struct {
+	client *dynamic.DynamicClient
+	scheme *runtime.Scheme
+}
+
+func NewEffectiveTLSPoliciesReconciler(client *dynamic.DynamicClient, scheme *runtime.Scheme) *EffectiveTLSPoliciesReconciler {
+	return &EffectiveTLSPoliciesReconciler{client: client, scheme: scheme}
+}
+
+func (t *EffectiveTLSPoliciesReconciler) Subscription() *controller.Subscription {
+	return &controller.Subscription{
+		Events: []controller.ResourceEventMatcher{
+			{Kind: &machinery.GatewayGroupKind},
+			{Kind: &kuadrantv1alpha1.TLSPolicyGroupKind},
+			{Kind: &CertManagerCertificateKind},
+		},
+		ReconcileFunc: t.Reconcile,
+	}
+}
+
+//+kubebuilder:rbac:groups=kuadrant.io,resources=tlspolicies,verbs=get;list;watch;update;patch;delete
+//+kubebuilder:rbac:groups=kuadrant.io,resources=tlspolicies/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=kuadrant.io,resources=tlspolicies/finalizers,verbs=update
+//+kubebuilder:rbac:groups="cert-manager.io",resources=issuers,verbs=get;list;watch;
+//+kubebuilder:rbac:groups="cert-manager.io",resources=clusterissuers,verbs=get;list;watch;
+//+kubebuilder:rbac:groups="",resources=secrets,verbs=get;list;watch
+//+kubebuilder:rbac:groups="cert-manager.io",resources=certificates,verbs=get;list;watch;create;update;patch;delete
+
+func (t *EffectiveTLSPoliciesReconciler) Reconcile(ctx context.Context, _ []controller.ResourceEvent, topology *machinery.Topology, _ error, s *sync.Map) error {
+	logger := controller.LoggerFromContext(ctx).WithName("EffectiveTLSPoliciesReconciler").WithName("Reconcile")
+
+	// Get all TLS Policies
+	policies := lo.Filter(topology.Policies().Items(), func(item machinery.Policy, index int) bool {
```

Comment:
So are we reconciling every policy on the cluster with every change? Not quite understanding how this is working

Reply:
That's fair. There's is a way to do reconcile only affected tls policies instead by using the events received. I've pushed a commit for this so this should be addressed now https://github.com/Kuadrant/kuadrant-operator/pull/927/commits/cd35907ece1813cb9f7ed7c70acbe092ebcbf52f

Reply:
I've reverted it since it looks to give more concerns that needed. Let's have a look can this be optimized in another PR or later point